Add LightStep tracing docs (#2844)
istio/istio.io: content/docs/tasks/telemetry/distributed-tracing-lightstep/index.md

> ---

Reviewer: can you convert the existing tracing task into a subsection thingy that has two tasks (with Zipkin, with Lightstep) ?
otherwise, this shows up as yet another task in the main index of tasks under networking

Reviewer 2: @geeknoid @rshriram we need some sort of consistent policy for vendor integration and documentation here, I think. Is our plan to provided "With X" sections for all policy/telemetry tasks? Or should we have an "Integrations" tab for tags for customizations, etc. ?

title: Distributed Tracing With LightStep [𝑥]PM
description: How to configure the proxies to send tracing requests to LightStep [𝑥]PM.
weight: 11
keywords: [telemetry,tracing,lightstep]
---

This task shows you how to configure Istio to collect trace spans and send them to LightStep [𝑥]PM.
[𝑥]PM lets you analyze 100% of unsampled transaction data from large-scale production software to produce meaningful
distributed traces and metrics that help explain performance behaviors and accelerate root cause analysis. For more

> information, visit [LightStep](https://lightstep.com).

Reviewer: :). I would kill this ad bit.. May be state it as "Lightstep is a hosted version of "run-your-own-zipkin" with a several value added features. For more info..."

        --set global.tracer.lightstep.accessToken="<LightStep Access Token>" \
        --set global.tracer.lightstep.secure=true \
        --set global.tracer.lightstep.cacertPath="/etc/lightstep/cacert.pem" \
        install/kubernetes/helm/istio \
        --name istio --namespace istio-system > $HOME/istio.yaml
    $ kubectl create namespace istio-system
    $ kubectl apply -f $HOME/istio.yaml
    {{< /text >}}

1.  Create a secret in the default namespace (or whichever namespace you are deploying the Bookinfo application into)
    with the Certificate Authority certificate used for verifying TLS communication with the Satellite pool.

    {{< text bash >}}
    $ CACERT=$(cat Cert_Auth.crt | base64) # Cert_Auth.crt contains the necessary CACert
    $ NAMESPACE=default
    {{< /text >}}

    ```bash
    $ cat <<EOF | kubectl apply -f -
      apiVersion: v1
      kind: Secret
      metadata:
        name: lightstep.cacert
        namespace: $NAMESPACE
        labels:
          app: lightstep
      type: Opaque
      data:
        cacert.pem: $CACERT
    EOF
    ```

1.   Deploy the [Bookinfo](/docs/examples/bookinfo/) sample application.

## Visualize trace data

> 1.  Send traffic to the sample application.

Author: I addressed this by removing the part about loading in the browser and sticking to `curl`.

    For the
    [Bookinfo](/docs/examples/bookinfo/#determining-the-ingress-ip-and-port)

>     sample, visit `http://$GATEWAY_URL/productpage` in your web browser

Reviewer 3: What do you mean with "For the Bookinfo sample"? The entire document uses the sample. Where can the reader obtain the value of `$GATEWAY_URL`? The variable is not mentioned anywhere else in this doc. Do not assume readers have covered other documents or followed their instructions to the letter.

Author: I addressed this by explicitly asking the user to follow the instructions at that link and added a step to confirm that the environment variable was set.

    or issue the following command:

    {{< text bash >}}

>     $ curl http://$GATEWAY_URL/productpage

Reviewer 3: If I run the command after following the commands in this document the command fails. Nowhere is the `$GATEWAY_URL` variable exported.

    {{< /text >}}

> 1.  Load the LightStep [𝑥]PM web UI then navigate to Explorer and find the query bar at the top. The query bar allows

Reviewer 3: The first sentence of this step has three different actions. In total, I count five different actions all of which should be in its own step.

    you to interactively filter results by a Service, Operation, and/or Tag values. Select `productpage.default` from

>     the Service drop-down list and click "Run". You should see something similar to the following:

Reviewer 3: Avoid using "You should" to avoid verb congruence issues. Opt instead for the simple present. "You see something ..."

{{< image width="100%" ratio="50%"
    link="./istio-tracing-list-lightstep.png"
    caption="Explorer"
    >}}

> If you click on the first row in the table of example traces below the latency histogram, you should see the details

Reviewer 3: This is another step. Please write it as such:
```suggestion
1. Click on the first row in the table of example traces below the latency histogram to see the details
```

> corresponding to your refresh of the `/productpage`. The page should look something like this:

Reviewer 3: Avoid conjunctives:
```suggestion
corresponding to your refresh of the `/productpage`. The page then looks similar to:
```

{{< image width="100%" ratio="50%"
    link="./istio-tracing-details-lightstep.png"
    caption="Detailed Trace View"
    >}}

> As you can see, the trace is comprised of a set of spans,

Reviewer 3: As I can see, where? Avoid using implicit referential language. Replace it with explicit references:
```suggestion
The screenshot shows that the trace is comprised of a set of spans.
```

> where each span corresponds to a Bookinfo service invoked during the execution of a `/productpage` request.

Reviewer 3: Run-on sentence:
```suggestion
Each span corresponds to a Bookinfo service invoked during the execution of a `/productpage` request.
```

> Every RPC is represented by two spans in the trace. For example, the call from `productpage` to `reviews` starts

Reviewer 3: Consider having this text at the beginning of the section instead of at the end. That way, readers have a better understanding of what they are doing and seeing as they follow the steps.
That said, the text has several issues which I address line by line. Passive voice:
```suggestion
Two spans in the trace represent every RPC. For example, the call from `productpage` to `reviews` starts
```

> with the span labeled with operation `reviews.default.svc.cluster.local:9080/*` and service

Reviewer 3: Prepositional issues:
```suggestion
with the span labeled with the `reviews.default.svc.cluster.local:9080/*` operation and the
```

> `productpage.default: proxy client`, which represents the client-side span for the call. It took 15.30 ms. The second

Reviewer 3: Will readers obtain the exact same response times? If you are referencing the values on the screenshots, then make the reference explicit to avoid confusion. The fix continues:
```suggestion
`productpage.default: proxy client` service, which represents the client-side span of the call. The screenshot shows that the call took 15.30 ms. The second
```

> span (labeled with operation `reviews.default.svc.cluster.local:9080/*` and service `reviews.default: proxy server`)

Reviewer 3: Do not use parenthesis; they are ambiguous in meaning.
```suggestion
span is labeled with the `reviews.default.svc.cluster.local:9080/*` operation and the `reviews.default: proxy server` service.
```

> is a child of the first span and represents the server-side span for the call. It took 14.60 ms.

Reviewer 3: Run-on sentence.
```suggestion
The second span is a child of the first span and represents the server-side span of the call. The screenshot shows the call took 14.60 ms.
```

> > {{< warning_icon >}} The LightStep integration does not currently capture spans generated by Istio's internal operation (ex. Mixer).

Reviewer 3: Mind the line length. Avoid parenthesis:
```suggestion
> {{< warning_icon >}} The LightStep integration does not currently capture spans generated by Istio's internal operation components such as Mixer.
```

> ## Understanding what happened

Reviewer 3: Please consider removing this entire section or at the very least change the title. Not only are you introducing new things like distributed tracing and Jaeger, but you also only provide a link for readers to find answers without any direct relation to the content of this document.

> To develop an understanding of how this trace span data is generated and then propagated, see the discussion in [Distributed Tracing With Jaeger](../distributed-tracing/#understanding-what-happened).

Reviewer 3: The language is not ideal. Better would be:
```suggestion
For more information on how this trace span data is generated and then propagated, visit the distributed tracing with Jaeger ["Understanding what happened" section](../distributed-tracing/#understanding-what-happened).
```
However, I believe this reference raises more questions than it answers.

## Trace sampling

> Istio captures a trace for all requests by default but you can lower the trace sampling percentage as needed.

Reviewer 3: Consider putting the context first:
```suggestion
By default, Istio captures a trace for all requests but you can lower the trace sampling percentage as needed.
```

> For information on how to do so, see the discussion in [Distributed Tracing With Jaeger](../distributed-tracing/#trace-sampling).

Reviewer 3: Please be explicit on what content is found in the cross-reference:
```suggestion
To learn how to reduce the trace sampling percentage, visit the [Distributed Tracing With Jaeger trace sampling section](../distributed-tracing/#trace-sampling).
```

> When using LightStep [𝑥]PM, reducing the trace sampling percentage is not recommended. In order to handle a high traffic mesh,

Reviewer 3: Passive voice:
```suggestion
When using LightStep [𝑥]PM, we do not recommend reducing the trace sampling percentage. To handle a high traffic mesh,
```

> consider scaling up the size of your Satellite pool.

Reviewer 3: Do not use capitalization for emphasis:
```suggestion
consider scaling up the size of your *satellite* pool.
```

## Cleanup

If you are not planning any follow-up tasks, remove the Bookinfo sample application and any LightStep [𝑥]PM secrets
from your cluster.

1. To remove the Bookinfo application, refer to the [Bookinfo cleanup](/docs/examples/bookinfo/#cleanup) instructions.

> 1. Remove the secret generated for LightStep [𝑥]PM

Reviewer 3: Punctuation:
```suggestion
1. Remove the secret generated for LightStep [𝑥]PM:
```

1.  Deploy Istio with the following configuration parameters specified:
    - `global.proxy.tracer="lightstep"`
    - `global.tracer.lightstep.address="<satellite-address>"`
    - `global.tracer.lightstep.accessToken="<access-token>"`
    - `global.tracer.lightstep.secure=true`
    - `global.tracer.lightstep.cacertPath="/etc/lightstep/cacert.pem"`

    If you are installing via `helm template` you can set these parameters using the `--set key=value` syntax
    when you run the `helm` command. For example:

    {{< text bash >}}
    $ helm template \
        --set global.proxy.tracer="lightstep" \
        --set global.tracer.lightstep.address="<satellite-address>" \
        --set global.tracer.lightstep.accessToken="<access-token>" \
        --set global.tracer.lightstep.secure=true \
        --set global.tracer.lightstep.cacertPath="/etc/lightstep/cacert.pem" \
        install/kubernetes/helm/istio \
        --name istio --namespace istio-system > $HOME/istio.yaml
    $ kubectl create namespace istio-system
    $ kubectl apply -f $HOME/istio.yaml
    {{< /text >}}

1.  Store your satellite pool's certificate authority certificate as a secret in the default namespace.

>     If the Bookinfo application will be deployed in a different namespace, create the secret in that namespace instead.

Reviewer 3: Verb congruence: No need to use the future tense or passive voice.
```suggestion
    If you deploy the Bookinfo application in a different namespace, create the secret in that namespace instead.
```

    {{< text bash >}}
    $ CACERT=$(cat Cert_Auth.crt | base64) # Cert_Auth.crt contains the necessary CACert
    $ NAMESPACE=default
    {{< /text >}}

    ```bash
    $ cat <<EOF | kubectl apply -f -
      apiVersion: v1
      kind: Secret
      metadata:
        name: lightstep.cacert
        namespace: $NAMESPACE
        labels:
          app: lightstep
      type: Opaque
      data:
        cacert.pem: $CACERT
    EOF
    ```

> 1.   Deploy the Bookinfo sample application by following [these instructions](/docs/examples/bookinfo/#deploying-the-application).

Reviewer 3: Use descriptive text for links. 
```suggestion
1.   Follow the [instructions to deploy the Bookinfo sample application](/docs/examples/bookinfo/#deploying-the-application).
```

## Visualize trace data

> 1.  Create an ingress gateway for the Bookinfo application by following

Reviewer 3: Use descriptive text for links.
```suggestion
1.  Follow the [instructions to create an ingress gateway for the Bookinfo application](/docs/examples/bookinfo/#determining-the-ingress-ip-and-port).
```

>     [these instructions](/docs/examples/bookinfo/#determining-the-ingress-ip-and-port).

Reviewer 3: ```suggestion
```

> 1.  Verify the previous step's success by confirming that the environment variable `GATEWAY_URL` is

Reviewer 3: Simplify the grammar to enhance clarity:
```suggestion
1.  To verify the previous step's success, confirm that you set the `GATEWAY_URL` environment variable in your shell.
```

>     populated in your shell.

Reviewer 3: ```suggestion
```

1.  Send traffic to the sample application.

    {{< text bash >}}
    $ curl http://$GATEWAY_URL/productpage
    {{< /text >}}

> 1.  Load the LightStep [𝑥]PM web UI.

Reviewer 3: I would add the link to the web UI here:
```suggestion
1.  Load the LightStep [𝑥]PM [web UI](<link.to.the.UI.com>).
```

1.  Navigate to Explorer.

> 1.  Find the query bar at the top. The query bar allows you to interactively filter results by a Service, Operation, and/or Tag values.

Reviewer 3: Do not use and/or, it introduces unnecessary ambiguity, I would also bold the labels for emphasis:
```suggestion
1.  Find the query bar at the top. The query bar allows you to interactively filter results by a **Service**, **Operation**, and **Tag** values.
```

> 1.  Select `productpage.default` from the Service drop-down list.

Reviewer 3: Bold the UI element:
```suggestion
1.  Select `productpage.default` from the **Service** drop-down list.
```

1.  Click **Run**. You see something similar to the following:

    {{< image width="100%" ratio="50%"
    link="./istio-tracing-list-lightstep.png"
    caption="Explorer"
    >}}

1.  Click on the first row in the table of example traces below the latency histogram to see the details
    corresponding to your refresh of the `/productpage`. The page then looks similar to:

    {{< image width="100%" ratio="50%"
    link="./istio-tracing-details-lightstep.png"
    caption="Detailed Trace View"
    >}}

The screenshot shows that the trace is comprised of a set of spans. Each span corresponds to a Bookinfo service invoked
during the execution of a `/productpage` request.

Two spans in the trace represent every RPC. For example, the call from `productpage` to `reviews` starts
with the span labeled with the `reviews.default.svc.cluster.local:9080/*` operation and the

> `productpage.default: proxy client` service, which represents the client-side span of the call. The screenshot shows

Reviewer 3: Ambiguity due to compound subject of the dependent clause. Basically, readers cannot determine whether it is the service, the operation, or both that represent the client-side of the call. The entire sentence, beginning in line 116 and ending here is a long complex run-on sentence. I suggest splitting it:
```suggestion
`productpage.default: proxy client` service. This service represents the client-side span of the call. The screenshot shows
```
